common/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,158 @@
+			Strategy: eventbusv1alpha1.AuthStrategyNone,
+		}
+	} else {
+		v := viper.New()
+		v.SetConfigName("auth")
+		v.SetConfigType("yaml")
+		v.AddConfigPath(common.EventBusAuthFileMountPath)
+		err := v.ReadInConfig()
+		if err != nil {
+			return nil, errors.Errorf("failed to load auth.yaml. err: %+v", err)
+		}
+		err = v.Unmarshal(cred)
+		if err != nil {
+			logger.Errorw("failed to unmarshal auth.yaml", zap.Error(err))
+			return nil, err
+		}
+		v.WatchConfig()
+		v.OnConfigChange(func(e fsnotify.Event) {
+			logger.Info("eventbus auth config file changed.")
+			err = v.Unmarshal(cred)
+			if err != nil {
+				logger.Errorw("failed to unmarshal auth.yaml after reloading", zap.Error(err))
+			}
+		})
+		auth = &eventbusdriver.Auth{
+			Strategy:    *eventBusAuth,
+			Crendential: cred,
+		}
+	}
+	var elector Elector
+	switch eventBusType {
+	case apicommon.EventBusNATS:
+		elector = &natsEventBusElector{
+			clusterName: clusterName,
+			size:        clusterSize,
+			url:         eventBusConfig.NATS.URL,
+			auth:        auth,
+		}
+	default:
+		return nil, errors.New("invalid eventbus type")
+	}
+	return elector, nil
+}
+
+type natsEventBusElector struct {
+	clusterName string
+	size        int
+	url         string
+	auth        *eventbusdriver.Auth
+}
+
+func (e *natsEventBusElector) RunOrDie(ctx context.Context, callbacks LeaderCallbacks) {
+	log := logging.FromContext(ctx)
+	ci := graft.ClusterInfo{Name: e.clusterName, Size: e.size}
+	opts := &nats.DefaultOptions
+	opts.Url = e.url
+	if e.auth.Strategy == eventbusv1alpha1.AuthStrategyToken {
+		opts.Token = e.auth.Crendential.Token
+	}
+	rpc, err := graft.NewNatsRpc(opts)
+	if err != nil {
+		log.Fatalw("failed to new Nats Rpc", zap.Error(err))
+	}
+	errChan := make(chan error)
+	stateChangeChan := make(chan graft.StateChange)
+	handler := graft.NewChanHandler(stateChangeChan, errChan)
+	node, err := graft.New(ci, handler, rpc, "/tmp/graft.log")
+	if err != nil {
+		log.Fatalw("failed to new a node", zap.Error(err))
+	}
+	defer node.Close()
+
+	cctx, cancel := context.WithCancel(ctx)
+	if node.State() == graft.LEADER {
+		log.Info("I'm the LEADER, starting ...")
+		go callbacks.OnStartedLeading(cctx)
+	} else {
+		log.Info("Not the LEADER, stand by ...")
+	}
+
+	handleStateChange := func(sc graft.StateChange) {
+		switch sc.To {
+		case graft.LEADER:
+			log.Info("I'm the LEADER, starting ...")
+			go callbacks.OnStartedLeading(cctx)
+		case graft.FOLLOWER, graft.CANDIDATE:
+			log.Infof("Becoming a %v, stand by ...", sc.To)
+			if sc.From == graft.LEADER {
+				cancel()
+				callbacks.OnStoppedLeading()
+				cctx, cancel = context.WithCancel(ctx)
+			}
+		case graft.CLOSED:
+			if sc.From == graft.LEADER {
+				cancel()
+				callbacks.OnStoppedLeading()
+			}
+			log.Fatal("Leader elector connection was CLOSED")
+		default:
+			log.Fatalf("Unknown state: %s", sc.To)
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("exiting...")
+			cancel()
```

Comment:
normally defer a cancel?

Reply:
When current node is changed from `leader` to non-leader (line 128-133), `cancel()` need to be called to terminate the running service, and re-initiate a cctx and cancel. Not quite sure if `defer cancel()` still works in that case, let me do more testing.

Reply:
Using defer works, updated.